hedera-node/hedera-consensus-service-impl/src/main/java/com/hedera/node/app/service/consensus/impl/handlers/customfee/ConsensusCustomFeeAssessor.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.consensus.impl.handlers.customfee;
+
+import static com.hedera.node.app.spi.workflows.HandleException.validateTrue;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.hedera.hapi.node.base.AccountAmount;
+import com.hedera.hapi.node.base.AccountID;
+import com.hedera.hapi.node.base.ResponseCodeEnum;
+import com.hedera.hapi.node.base.TokenID;
+import com.hedera.hapi.node.base.TokenTransferList;
+import com.hedera.hapi.node.base.TransferList;
+import com.hedera.hapi.node.state.consensus.Topic;
+import com.hedera.hapi.node.token.CryptoTransferTransactionBody;
+import com.hedera.hapi.node.transaction.ConsensusCustomFee;
+import com.hedera.hapi.node.transaction.FixedFee;
+import com.hedera.node.app.service.token.ReadableTokenStore;
+import com.hedera.node.app.spi.workflows.HandleContext;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.List;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+@Singleton
+public class ConsensusCustomFeeAssessor {
+
+    /**
+     * Constructs a {@link ConsensusCustomFeeAssessor} instance.
+     */
+    @Inject
+    public ConsensusCustomFeeAssessor() {
+        // Needed for Dagger injection
+    }
+
+    /**
+     * Build and return a list of synthetic crypto transfer transaction bodies, that represents custom fees payments.
+     * It will return one body per topic custom fee.
+     *
+     * @param topic The topic
+     * @param context The transaction handle context
+     * @return List of synthetic crypto transfer transaction bodies
+     */
+    public List<CryptoTransferTransactionBody> assessCustomFee(
+            @NonNull final Topic topic, @NonNull final HandleContext context) {
+        final List<CryptoTransferTransactionBody> transactionBodies = new ArrayList<>();
```

Comment:
I think it will be easier to pass in the maps constructed to create the synthetic bodies

Reply:
I'm not sure if I understand you correctly, but the maps from the previous comment don't contain the necessary information in order to build the crypto transfer bodies. 

Anyway, I did some refactoring. 
Now, before validation of the `max_custom_fees` and synthetic body creations, we filter the list of custom fees and use only chargeable fees. (we will skip the fees where the payer is collector or token treasury)
This way we don't need to check twice, if a given fee should be validated or to create a body with it.